iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/compaction/IcebergCompactionService.java

```diff
@@ -63,6 +59,16 @@ public Boolean compact(Table table, CompactionInfo ci) throws Exception {
       return false;
     }
 
+    if (ci.type == CompactionType.SMART_OPTIMIZE) {
+      CompactionType type = compactionEvaluator.determineCompactionType();
+      if (type == null) {
+        msc.markRefused(CompactionInfo.compactionInfoToStruct(ci));
+        return false;
+      }
+      ci.type = type;
+      msc.updateCompactorState(CompactionInfo.compactionInfoToStruct(ci), -1);
```

Comment:
should we call it only inside of `SMART_OPTIMIZE`? don't we need to update `runAs` as well?

Reply:
Fixed, moved `msc.updateCompactorState` to be always called.